### PR TITLE
Specific mutation descriptions

### PR DESCRIPTION
**Pull request by nhaajt**
## Overview
+ Update mutator names, descriptions, and docs for better consistency
+ Add the ability to define better descriptions specific to each mutation
+ Add specific mutation descriptions for built-in mutators

## Specific mutation description
By default, it is the same as the description of the mutator with the addition of the start position of the mutation. For example:
```scala
// Before
"Remove a beginning of line character `^`"
// After
"Remove a beginning of line character `^` at 0:0"
```

Specific mutation description can be defined per mutator that extends `TokenMutator` by overriding the `describeMutation` method, which provides the token being mutated and the mutation location:
```scala
override def describeMutation(token: RegexTree, location: Location): String =
 s "The token ${token} is being mutated at ${location}"
```

New mutation description example for `CharClassChildRemoval` mutator:
```scala
// Original regex
"[ab0-9[A-Z][cd]]"
// Mutants
"[b0-9[A-Z][cd]]"
"[a0-9[A-Z][cd]]"
"[ab[A-Z][cd]]"
"[ab0-9[cd]]"
"[ab0-9[A-Z]]"
"[ab0-9[A-Z][d]]"
"[ab0-9[A-Z][c]]"

// Before (same for all mutants)
"Remove a child from a character class"
// After
"Remove the child Character at [0:1, 0:2) from the character class"
"Remove the child Character at [0:2, 0:3) from the character class"
"Remove the child Range at [0:3, 0:6) from the character class"
"Remove the child CharacterClass at [0:6, 0:11) from the character class"
"Remove the child CharacterClass at [0:11, 0:15) from the character class"
"Remove the child Character at [0:12, 0:13) from the character class"
"Remove the child Character at [0:13, 0:14) from the character class"
```

### CODE REVIEW

File: core/src/main/scala/weaponregex/model/mutation/TokenMutator.scala [hunk 1]
```diff
@@ -18,6 +19,18 @@ trait TokenMutator {
...
+    * the description of the mutator with the addition of the start position from the given location.
+    *
+    * @param token
+    *   The token to be mutated
+    * @param location
+    *   The location of the mutation
+    * @return
+    *   A description of the mutation
+    */
+  def describeMutation(@nowarn token: RegexTree, location: Location): String = s"$description at ${location.start}"
```
Comment (hugo-vrijswijk) [inline core/src/main/scala/weaponregex/model/mutation/TokenMutator.scala:32]: I'm not too sure about this pattern. It seems like it's doing almost the same thing as the `description` method. It also seems inconsistent in adding the location at the end for some mutations, and a different description for some others.

File: build.sbt [hunk 2]
```diff
@@ -47,7 +47,8 @@ lazy val WeaponRegeX = projectMatrix
     name := "weapon-regex",
     libraryDependencies += "com.lihaoyi" %%% "fastparse" % "2.3.3",
     libraryDependencies += "org.scalameta" %%% "munit" % "0.7.29" % Test,
-    tpolecatScalacOptions += ScalacOption("-Xsource:3", _.major == 2)
+    tpolecatScalacOptions += ScalacOption("-Xsource:3", _.major == 2),
+    tpolecatScalacOptions += ScalacOption("-Xlint:unused", _.major == 2)
```
Comment (hugo-vrijswijk) [inline build.sbt:51]: This option is already enabled, I think: 

https://docs.scala-lang.org/overviews/compiler-options/index.html

image

image

File: core/src/main/scala/weaponregex/mutator/capturingMutator.scala [hunk 3]
```diff
@@ -13,24 +14,29 @@ import weaponregex.model.regextree.*
...
   *   `(?=abc)` ⟶ `(?!abc)`
   */
 object LookaroundNegation extends TokenMutator {
-  override val name: String = "Lookaround constructs negation"
+  override val name: String = "Lookaround construct (lookahead, lookbehind) negation"
   override val levels: Seq[Int] = Seq(1, 2, 3)
-  override val description: String = "Negate lookaround constructs (lookahead, lookbehind)"
+  override val description: String = "Negate a lookaround construct (lookahead, lookbehind)"
+
+  override def describeMutation(token: RegexTree, location: Location): String = token match {
```
[the remark this answers is not in the document]
Comment (Nhaajt) [reply core/src/main/scala/weaponregex/mutator/capturingMutator.scala:36]: That’s what I initially wanted, but I tried not to make breaking changes. You can think of `description` as the description for the mutator, while the new method `describeMutation` describes a specific mutation.

File: core/src/main/scala/weaponregex/model/mutation/TokenMutator.scala (hunk 1 above)
Comment (Nhaajt) [reply core/src/main/scala/weaponregex/model/mutation/TokenMutator.scala:32]: I’m also still puzzled about the description pattern consistency. I’m thinking of always having the location first, but it reads better like this, in my opinion. Any suggestions?
```scala
"Remove the child Character at [0:1, 0:2) from the character class"
"[0:1, 0:2) Remove a child Character from the character class"
```

File: build.sbt (hunk 2 above)
Comment (Nhaajt) [reply build.sbt:51]: Strange 🤔 I got [this error](https://github.com/stryker-mutator/weapon-regex/actions/runs/3350979650/jobs/5552131281) `@nowarn annotation does not suppress any warnings`, which seems like what is described [here](https://www.scala-lang.org/2021/01/12/configuring-and-suppressing-warnings.html#:~:text=To%20ensure%20that%20%40nowarn%20annotations%20actually%20suppress%20warnings%2C%20enable%20%2DXlint%3Aunused%20or%20%2DWunused%3Anowarn.%20With%20this%20option%2C%20the%20compiler%20checks%20that%20every%20%40nowarn%20annotation%20suppresses%20at%20least%20one%20warning%20and%20issues%20a%20warning%20otherwise%3A).

Comment (hugo-vrijswijk) [reply build.sbt:51]: Stryker4s disables some compiler options to prevent unrelated compiler errors, that might be causing it

Comment (Nhaajt) [reply build.sbt:51]: Then how should I resolve this?

**Review by hugo-vrijswijk — APPROVED** [review]
@nhaajt  Sorry this has also been open for so long. I think it looks good. You can merge if you want

**Comment by nhaajt** [comment]
A (hopefully) better implementation of an even more detailed mutant description is now available on [feat/better-mutant-description](https://github.com/stryker-mutator/weapon-regex/tree/feat/better-mutant-description) branch.

**Pull request closed without merging**